### PR TITLE
fix some resolutions in create-app template

### LINKED ISSUE #22142: 🐛 Bug Report: "Getting Started" failing
### 📜 Description

Simple app, following [Getting Started](https://backstage.io/docs/getting-started/) Guide fails.

### 👍 Expected behavior

Simple app should just run.

### 👎 Actual Behavior with Screenshots

A compilation failure, possibly related to Rect, occurs, with the following error:

```
[0] ERROR in ../../node_modules/swagger-ui-react/node_modules/react-redux/dist/react-redux.mjs 1097:18-45
[0] export 'useSyncExternalStore' (imported as 'React2') was not found in 'react' (possible exports: Children, Component, Fragment, Profiler, PureComponent, StrictMode, Suspense, __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, cloneElement, createContext, createElement, createFactory, createRef, forwardRef, isValidElement, lazy, memo, useCallback, useContext, useDebugValue, useEffect, useImperativeHandle, useLayoutEffect, useMemo, useReducer, useRef, useState, version)
[0]
[0] webpack compiled with 1 error
```

### 👟 Reproduction steps

To avoid local environment inconsistencies, I've managed to reproduce the bug with the following steps, using a Docker container:

1. Run `docker run -it --rm -p 3000:3000 -p 7007:7007 node:20 bash`
2. Run `npx @backstage/create-app@latest` (all default answers)
3. Run `cd backstage` and then `yarn dev`

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

Problem occurs both in my local machine and in a docker container. Here:

**Local machine**:
```
yarn run v1.22.19
$ /Users/pedro/Developer/GET-INSIGHT/RNP/Arquitetura/PoCs/tmp/backstage/node_modules/.bin/backstage-cli info
OS:   Darwin 23.2.0 - darwin/arm64
node: v18.16.1
yarn: 1.22.19
cli:  0.25.0 (installed)
backstage:  1.21.0

Dependencies:
  @backstage/app-defaults                                          1.4.6
  @backstage/backend-app-api                                       0.5.9
  @backstage/backend-common                                        0.20.0
  @backstage/backend-dev-utils                                     0.1.2
  @backstage/backend-openapi-utils                                 0.1.1
  @backstage/backend-plugin-api                                    0.6.8
  @backstage/backend-tasks                                         0.5.13
  @backstage/catalog-client                                        1.5.1
  @backstage/catalog-model                                         1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.1
  @backstage/cli                                                   0.25.0
  @backstage/config-loader                                         1.6.0
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.2
  @backstage/core-compat-api                                       0.1.0
  @backstage/core-components                                       0.13.9
  @backstage/core-plugin-api                                       1.8.1
  @backstage/e2e-test-utils                                        0.1.0
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.4
  @backstage/frontend-plugin-api                                   0.4.0
  @backstage/integration-aws-node                                  0.1.8
  @backstage/integration-react                                     1.1.22
  @backstage/integration                                           1.8.0
  @backstage/plugin-api-docs                                       0.10.2
  @backstage/plugin-app-backend                                    0.3.56
  @backstage/plugin-app-node                                       0.1.8
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.2
  @backstage/plugin-auth-backend-module-github-provider            0.1.5
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.5
  @backstage/plugin-auth-backend-module-google-provider            0.1.5
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.5
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.0
  @backstage/plugin-auth-backend-module-okta-provider              0.0.1
  @backstage/plugin-auth-backend                                   0.20.2
  @backstage/plugin-auth-node                                      0.4.2
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.5
  @backstage/plugin-catalog-backend                                1.16.0
  @backstage/plugin-catalog-common                                 1.0.19
  @backstage/plugin-catalog-graph                                  0.3.2
  @backstage/plugin-catalog-import                                 0.10.4
  @backstage/plugin-catalog-node                                   1.6.0
  @backstage/plugin-catalog-react                                  1.9.2
  @backstage/plugin-catalog                                        1.16.0
  @backstage/plugin-events-node                                    0.2.17
  @backstage/plugin-github-actions                                 0.6.9
  @backstage/plugin-org                                            0.6.18
  @backstage/plugin-permission-common                              0.7.11
  @backstage/plugin-permission-node                                0.7.19
  @backstage/plugin-permission-react                               0.4.18
  @backstage/plugin-proxy-backend                                  0.4.6
  @backstage/plugin-scaffolder-backend-module-azure                0.1.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.1.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.0
  @backstage/plugin-scaffolder-backend-module-github               0.1.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.11
  @backstage/plugin-scaffolder-backend                             1.19.2
  @backstage/plugin-scaffolder-common                              1.4.4
  @backstage/plugin-scaffolder-node                                0.2.9
  @backstage/plugin-scaffolder-react                               1.7.0
  @backstage/plugin-scaffolder                                     1.17.0
  @backstage/plugin-search-backend-module-catalog                  0.1.12
  @backstage/plugin-search-backend-module-pg                       0.5.17
  @backstage/plugin-search-backend-module-techdocs                 0.1.12
  @backstage/plugin-search-backend-node                            1.2.12
  @backstage/plugin-search-backend                                 1.4.8
  @backstage/plugin-search-common                                  1.2.9
  @backstage/plugin-search-react                                   1.7.4
  @backstage/plugin-search                                         1.4.4
  @backstage/plugin-tech-radar                                     0.6.11
  @backstage/plugin-techdocs-backend                               1.9.1
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.3
  @backstage/plugin-techdocs-node                                  1.11.0
  @backstage/plugin-techdocs-react                                 1.1.14
  @backstage/plugin-techdocs                                       1.9.2
  @backstage/plugin-user-settings                                  0.7.14
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.4.6
  @backstage/theme                                                 0.5.0
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7
Done in 0.83s.
```

*Docker container:*
```
$ /backstage/node_modules/.bin/backstage-cli info
OS:   Linux 6.4.16-linuxkit - linux/arm64
node: v20.10.0
yarn: 1.22.19
cli:  0.25.0 (installed)
backstage:  1.21.0

Dependencies:
  @backstage/app-defaults                                          1.4.6
  @backstage/backend-app-api                                       0.5.9
  @backstage/backend-common                                        0.20.0
  @backstage/backend-dev-utils                                     0.1.2
  @backstage/backend-openapi-utils                                 0.1.1
  @backstage/backend-plugin-api                                    0.6.8
  @backstage/backend-tasks                                         0.5.13
  @backstage/catalog-client                                        1.5.1
  @backstage/catalog-model                                         1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.2.1
  @backstage/cli                                                   0.25.0
  @backstage/config-loader                                         1.6.0
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.2
  @backstage/core-compat-api                                       0.1.0
  @backstage/core-components                                       0.13.9
  @backstage/core-plugin-api                                       1.8.1
  @backstage/e2e-test-utils                                        0.1.0
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.4
  @backstage/frontend-plugin-api                                   0.4.0
  @backstage/integration-aws-node                                  0.1.8
  @backstage/integration-react                                     1.1.22
  @backstage/integration                                           1.8.0
  @backstage/plugin-api-docs                                       0.10.2
  @backstage/plugin-app-backend                                    0.3.56
  @backstage/plugin-app-node                                       0.1.8
  @backstage/plugin-auth-backend-module-atlassian-provider         0.1.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.2
  @backstage/plugin-auth-backend-module-github-provider            0.1.5
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.5
  @backstage/plugin-auth-backend-module-google-provider            0.1.5
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.5
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.0
  @backstage/plugin-auth-backend-module-okta-provider              0.0.1
  @backstage/plugin-auth-backend                                   0.20.2
  @backstage/plugin-auth-node                                      0.4.2
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.5
  @backstage/plugin-catalog-backend                                1.16.0
  @backstage/plugin-catalog-common                                 1.0.19
  @backstage/plugin-catalog-graph                                  0.3.2
  @backstage/plugin-catalog-import                                 0.10.4
  @backstage/plugin-catalog-node                                   1.6.0
  @backstage/plugin-catalog-react                                  1.9.2
  @backstage/plugin-catalog                                        1.16.0
  @backstage/plugin-events-node                                    0.2.17
  @backstage/plugin-github-actions                                 0.6.9
  @backstage/plugin-org                                            0.6.18
  @backstage/plugin-permission-common                              0.7.11
  @backstage/plugin-permission-node                                0.7.19
  @backstage/plugin-permission-react                               0.4.18
  @backstage/plugin-proxy-backend                                  0.4.6
  @backstage/plugin-scaffolder-backend-module-azure                0.1.0
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.1.0
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.0
  @backstage/plugin-scaffolder-backend-module-github               0.1.0
  @backstage/plugin-scaffolder-backend-module-gitlab               0.2.11
  @backstage/plugin-scaffolder-backend                             1.19.2
  @backstage/plugin-scaffolder-common                              1.4.4
  @backstage/plugin-scaffolder-node                                0.2.9
  @backstage/plugin-scaffolder-react                               1.7.0
  @backstage/plugin-scaffolder                                     1.17.0
  @backstage/plugin-search-backend-module-catalog                  0.1.12
  @backstage/plugin-search-backend-module-pg                       0.5.17
  @backstage/plugin-search-backend-module-techdocs                 0.1.12
  @backstage/plugin-search-backend-node                            1.2.12
  @backstage/plugin-search-backend                                 1.4.8
  @backstage/plugin-search-common                                  1.2.9
  @backstage/plugin-search-react                                   1.7.4
  @backstage/plugin-search                                         1.4.4
  @backstage/plugin-tech-radar                                     0.6.11
  @backstage/plugin-techdocs-backend                               1.9.1
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.3
  @backstage/plugin-techdocs-node                                  1.11.0
  @backstage/plugin-techdocs-react                                 1.1.14
  @backstage/plugin-techdocs                                       1.9.2
  @backstage/plugin-user-settings                                  0.7.14
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.4.6
  @backstage/theme                                                 0.5.0
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7
Done in 0.47s.
```

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

### PR DESCRIPTION
**Pull request by jamieklassen**
## Hey, I just made a Pull Request!

The point of this is to resolve those annoying 'E2E Linux' checks that have been blocking everybody's PRs (9 of them from my team at the moment lol)!

I'm not confident about the decisions underlying #12252 or the answer to @Pike's question [here](https://github.com/backstage/backstage/issues/22142#issuecomment-1888991931), but I wanted to see this PR run through the CI checks and hopefully prompt some discussion about what kinds of resolutions are suitable for these templates. Each commit contains a bit of justification, but summarized here:
* I used the react 18 type definitions, which seemed to work around mui/material-ui#40427 and fix `yarn tsc:full` on a fresh app
* I added some `swagger-ui-react`-specific resolutions following the advice in [their wiki](https://github.com/swagger-api/swagger-ui/wiki/Using-older-version-of-React); this is an option to resolve #22142

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] ~Added or updated documentation~ N/A
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] ~Screenshots attached (for UI changes)~ no UI
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — CHANGES_REQUESTED** [review]
Resolutions are tricky : /

They don't apply to existing projects, and it is best to remove them when they are no longer needed, so they cause a lot of manual updates. They can also mess up existing projects that try to apply them with older dependencies or other resolutions in place. So far we've aimed to use them at a minimum and instead aim to have issues fixed upstream instead. The `@types/react` is just one of those where it actually makes sense to have a resolutions in place to select the version of React used in the repo.

File: packages/create-app/templates/default-app/package.json.hbs [hunk 1]
```diff
@@ -42,8 +42,11 @@
     "typescript": "~5.2.0"
   },
   "resolutions": {
-    "@types/react": "^17",
-    "@types/react-dom": "^17"
+    "@types/react": "^18",
+    "@types/react-dom": "^18",
```
[the remark this answers is not in the document]
Comment (jamieklassen) [reply packages/create-app/templates/default-app/package.json.hbs:46]: When you say "...along with any test utils", do you mean the template app should use test utils from React 18, or any Backstage test utils should use React 18?

Intuitively this seems like a big enough undertaking that I won't be able to fix it all in the scope of this PR; are these remaining "React 18" tasks ("update `src/index.tsx`", "update dependency ranges", etc) tracked somewhere? This [React 18 milestone](https://github.com/backstage/backstage/milestone/41) looks to be complete, but I also see comments like [this](https://github.com/backstage/backstage/issues/18433#issuecomment-1612669974) suggesting that React 18 and MUI 5 are a bit intertwined.

If it helps get PRs unblocked, I'd probably want to contribute to such an initiative!

File: packages/create-app/templates/default-app/package.json.hbs [hunk 2]
```diff
@@ -42,8 +42,11 @@
...
+    "swagger-ui-react/react": "^18",
+    "swagger-ui-react/react-dom": "^18",
+    "swagger-ui-react/react-redux": "^8"
```
[the remark this answers is not in the document]
Comment (jamieklassen) [reply packages/create-app/templates/default-app/package.json.hbs:49]: Yeah this was a bit of a blind effort to get around #22142. Re-reading the wiki article [here](https://github.com/swagger-api/swagger-ui/wiki/Using-older-version-of-React) it looks like this should not be required for React 18, provided we use a new enough version of `swagger-ui-react`.

File: packages/create-app/templates/default-app/package.json.hbs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jamieklassen) [reply packages/create-app/templates/default-app/package.json.hbs:46]: Hey, looks like you already handled it with #22297; nice!

**Comment by jamieklassen** [comment]
Redundant to #22297

File: packages/create-app/templates/default-app/package.json.hbs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/create-app/templates/default-app/package.json.hbs:46]: Yep, those where the missing bits 😁 - in particular the migration guide

**Pull request closed without merging**